### ISSUE TITLE
Enhancement: Allow disabling openapi documentation sites and schema download endpoints via config

### ISSUE THREAD

**Issue by Goldziher** [starlite-api/starlite#418]
Currently the OpenAPIController always exposes `redoc`, `swagger-ui` and `stoplight elements` static sites, and a `yaml` + `json` schema download endpoints. This though might not be desirable. 

Tasks:

1. add a new method on the controller called `render_404_not_found` or something such like, which renders a simple 404 page.
2. add a configuration option to the `OpenAPIConfig` class that determines which endpoints are enabled. For example, a list or set of literal identifiers - which by default includes all values. 
3. add logic to return a 404 if a particular endpoint is not enabled in the config.

Note:

Access the app level config is possble using the request instance. See how this is currently implemented in the `OpenAPIController.root` method as an example.

**Comment by ingjavierpinilla** [comment]
Writing a method that returns an html string would make the tests green even if the endpoint is not authorized because you would still have a HTTP 200 code.

Any idea how this could be solved? 
I thought about using the class ExceptionResponseContent but as far as I understood, it doesn't allow you to render html.

**Comment by Goldziher** [comment]
You can return the following:

```python
from starlite import Get, MediaType, Response
from starlette.status import HTTP_404_NOT_FOUND

@get(...)
def handler() -> Response:
   return Response(content=<html string>, status_code=HTTP_404_NOT_FOUND, media_type=MediaType.HTML)
```

**Comment by ingjavierpinilla** [comment]
Thanks for your answer.

Not completely, the to_response function treats the response as flat data and tries to construct a new response using a response object as the content parameter.

Or am I missing something? 
Is there any way to avoid this behavior and return the response directly?

Maybe it is an issue we want to address before this (418) one?

**Comment by Goldziher** [comment]
are you sure you are on an updated version of `starlite`? Look here `tests/handlers/http/test_to_response.py` - this is working as expected.

**Comment by RSPersonal** [comment]
Hi can i pick this up?

**Comment by ingjavierpinilla** [comment]
Sure

**Comment by ingjavierpinilla** [comment]
> Hi can i pick this up?

Sure @RSPersonal

**Comment by RSPersonal** [comment]
So bear with me, its my first issue on a open source project. I also seem to run into the same issue as @ingjavierpinilla. 

I wrote this method in the controller. 

` @staticmethod
    def render_404_not_found(self):
        html_response = f"""
            <!DOCTYPE html>
            <html lang="en">
            <head>
                <title>404</title>
                {self.favicon}
                <meta charset="utf-8"/>
                <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"/>
                <meta name="keywords" content=""/>
                <link href="https://fonts.googleapis.com/css?family=Montserrat:300,400,700|Roboto:300,400,700" rel="stylesheet">
            </head>
            <body>
            <header>
                <h2>Test</h2>
            </header>
            </body>
            </html>
        """
        return Response(content=html_response, status_code=HTTP_404_NOT_FOUND, media_type=MediaType.HTML)`

Even tho i pass the 404 in the status code, the sever still treats it as an 200 request. 

![image](https://user-images.githubusercontent.com/74533741/187131366-d737b01f-d44f-406b-aa3d-663a431a90c1.png)

Also the page does not render the html.
![image](https://user-images.githubusercontent.com/74533741/187131520-1cb25b17-4a46-4f68-a1a8-6b14b8b5361b.png)

**Comment by peterschutt** [comment]
Hey @RSPersonal - thanks for looking at this.

How about raising [NotFoundException`](https://starlite-api.github.io/starlite/reference/exceptions/#starlite.exceptions.NotFoundException) and register a handler for that on the [controller](https://starlite-api.github.io/starlite/reference/controller/#starlite.controller.Controller.exception_handlers) that returns the html response?

**Comment by RSPersonal** [comment]
alright, will look at it after work and update you

**Comment by Goldziher** [comment]
> alright, will look at it after work and update you

It would be beneficial to have a PR with your existing code. I'd like to understand what is going on there, because it should be working.

**Comment by peterschutt** [comment]
![image](https://user-images.githubusercontent.com/20659309/187165028-cdf2df7f-0925-42bd-9f6e-a29758dfc8db.png)

Looks like the route is returning `None`

**Comment by Goldziher** [comment]
> ![image](https://user-images.githubusercontent.com/20659309/187165028-cdf2df7f-0925-42bd-9f6e-a29758dfc8db.png)
> 
> Looks like the route is returning `None` - no `return`

Well then

**Comment by ingjavierpinilla** [comment]
> Hey @RSPersonal - thanks for looking at this.
> 
> How about raising [`NotFoundException`](https://starlite-api.github.io/starlite/reference/exceptions/#starlite.exceptions.NotFoundException) and register a handler for that on the [`Controller`](https://starlite-api.github.io/starlite/reference/controller/#starlite.controller.Controller.exception_handlers) that returns the html response?

Hi,
this works but wouldn't PermissionDeniedException be a better response?

**Comment by peterschutt** [comment]
I think 404 is the correct response - it should be as if the route doesn't exist.

Instead of the exception handler approach, it might be better to register a [`before_request`](https://starlite-api.github.io/starlite/reference/controller/#starlite.controller.Controller.before_request) hook on the controller that would intercept the request before it reaches the handler, inspect the request path to ascertain if the endpoint should be served or not and if not, return the 404 response object. I think that would better allow you to centralize the logic.

**Comment by Goldziher** [comment]
> I think 404 is the correct response - it should be as if the route doesn't exist.
> 
> Instead of the exception handler approach, it might be better to register a [`before_request`](https://starlite-api.github.io/starlite/reference/controller/#starlite.controller.Controller.before_request) hook on the controller that would intercept the request before it reaches the handler, inspect the request path to ascertain if the endpoint should be served or not and if not, return the 404 response object. I think that would better allow you to centralize the logic.

But we want to make the 404 template customizable

**Comment by ingjavierpinilla** [comment]
> I think 404 is the correct response - it should be as if the route doesn't exist.
> 
> Instead of the exception handler approach, it might be better to register a [`before_request`](https://starlite-api.github.io/starlite/reference/controller/#starlite.controller.Controller.before_request) hook on the controller that would intercept the request before it reaches the handler, inspect the request path to ascertain if the endpoint should be served or not and if not, return the 404 response object. I think that would better allow you to centralize the logic.

I thought about that too, but I didn't see an easy and clean way to check the path.

You propose using a before_request hook -> raise a NotFoundException -> use an exception_handler to catch the exception?

**Comment by peterschutt** [comment]
> > I think 404 is the correct response - it should be as if the route doesn't exist.
> > Instead of the exception handler approach, it might be better to register a [`before_request`](https://starlite-api.github.io/starlite/reference/controller/#starlite.controller.Controller.before_request) hook on the controller that would intercept the request before it reaches the handler, inspect the request path to ascertain if the endpoint should be served or not and if not, return the 404 response object. I think that would better allow you to centralize the logic.
> 
> But we want to make the 404 template customizable

Sorry, missed that spec - disregard the above as I agree in that case the method approach is the way to go.

Perhaps we should be returning response objects from the routes instead of `str` then:

```python
    @get(path="/swagger", include_in_schema=False)
    def swagger_ui(self, request: Request) -> Response:
        ...
```

**Comment by Goldziher** [comment]
> > > I think 404 is the correct response - it should be as if the route doesn't exist.
> > > Instead of the exception handler approach, it might be better to register a [`before_request`](https://starlite-api.github.io/starlite/reference/controller/#starlite.controller.Controller.before_request) hook on the controller that would intercept the request before it reaches the handler, inspect the request path to ascertain if the endpoint should be served or not and if not, return the 404 response object. I think that would better allow you to centralize the logic.
> > 
> > But we want to make the 404 template customizable
> 
> Sorry, missed that spec - disregard the above as I agree in that case the method approach is the way to go.
> 
> Perhaps we should be returning response objects from the routes instead of `str` then:
> 
> ```python
>     @get(path="/swagger", include_in_schema=False)
>     def swagger_ui(self, request: Request) -> Response:
>         ...
> ```

What would be the advantage?

**Comment by peterschutt** [comment]
> > I think 404 is the correct response - it should be as if the route doesn't exist.
> > Instead of the exception handler approach, it might be better to register a [`before_request`](https://starlite-api.github.io/starlite/reference/controller/#starlite.controller.Controller.before_request) hook on the controller that would intercept the request before it reaches the handler, inspect the request path to ascertain if the endpoint should be served or not and if not, return the 404 response object. I think that would better allow you to centralize the logic.
> 
> I thought about that too, but I didn't see an easy and clean way to check the path.
> 
> You propose using a before_request hook -> raise a NotFoundException -> use an exception_handler to catch the exception?

You wouldn't need the exception handler when using the `before_request` approach, you'd just return the 404 response object directly from the hook. But see @Goldziher's comment above - that approach wouldn't allow the 404 template to be easily customized.

**Comment by peterschutt** [comment]
> > > > I think 404 is the correct response - it should be as if the route doesn't exist.
> > > > Instead of the exception handler approach, it might be better to register a [`before_request`](https://starlite-api.github.io/starlite/reference/controller/#starlite.controller.Controller.before_request) hook on the controller that would intercept the request before it reaches the handler, inspect the request path to ascertain if the endpoint should be served or not and if not, return the 404 response object. I think that would better allow you to centralize the logic.
> > > 
> > > 
> > > But we want to make the 404 template customizable
> > 
> > 
> > Sorry, missed that spec - disregard the above as I agree in that case the method approach is the way to go.
> > Perhaps we should be returning response objects from the routes instead of `str` then:
> > ```python
> >     @get(path="/swagger", include_in_schema=False)
> >     def swagger_ui(self, request: Request) -> Response:
> >         ...
> > ```
> 
> What would be the advantage?

Just consistency - otherwise the route handlers would have to return a string for the 200 response, or a `Response` object in the case of the 404 to override the status code.. not that big a deal.

**Comment by Goldziher** [comment]
> > > > > I think 404 is the correct response - it should be as if the route doesn't exist.
> > > > > Instead of the exception handler approach, it might be better to register a [`before_request`](https://starlite-api.github.io/starlite/reference/controller/#starlite.controller.Controller.before_request) hook on the controller that would intercept the request before it reaches the handler, inspect the request path to ascertain if the endpoint should be served or not and if not, return the 404 response object. I think that would better allow you to centralize the logic.
> > > > 
> > > > 
> > > > But we want to make the 404 template customizable
> > > 
> > > 
> > > Sorry, missed that spec - disregard the above as I agree in that case the method approach is the way to go.
> > > Perhaps we should be returning response objects from the routes instead of `str` then:
> > > ```python
> > >     @get(path="/swagger", include_in_schema=False)
> > >     def swagger_ui(self, request: Request) -> Response:
> > >         ...
> > > ```
> > 
> > What would be the advantage?
> 
> Just consistency - otherwise the route handlers would have to return a string for the 200 response, or a `Response` object in the case of the 404 to override the status code.. not that big a deal.

Fair point . Can you show @RSPersonal how this would look?

**Comment by peterschutt** [comment]
> but I didn't see an easy and clean way to check the path.

Don't forget that the `Request` object extends [`starlette.Request`](https://www.starlette.io/requests/) - so you have `request.url.path` for example. We need to work out how best to get those starlette attributes to show in our ref. docs.

**Comment by peterschutt** [comment]
> > > > > > I think 404 is the correct response - it should be as if the route doesn't exist.
> > > > > > Instead of the exception handler approach, it might be better to register a [`before_request`](https://starlite-api.github.io/starlite/reference/controller/#starlite.controller.Controller.before_request) hook on the controller that would intercept the request before it reaches the handler, inspect the request path to ascertain if the endpoint should be served or not and if not, return the 404 response object. I think that would better allow you to centralize the logic.
> > > > > 
> > > > > 
> > > > > But we want to make the 404 template customizable
> > > > 
> > > > 
> > > > Sorry, missed that spec - disregard the above as I agree in that case the method approach is the way to go.
> > > > Perhaps we should be returning response objects from the routes instead of `str` then:
> > > > ```python
> > > >     @get(path="/swagger", include_in_schema=False)
> > > >     def swagger_ui(self, request: Request) -> Response:
> > > >         ...
> > > > ```
> > > 
> > > 
> > > What would be the advantage?
> > 
> > 
> > Just consistency - otherwise the route handlers would have to return a string for the 200 response, or a `Response` object in the case of the 404 to override the status code.. not that big a deal.
> 
> Fair point . Can you show @RSPersonal how this would look?

@RSPersonal @ingjavierpinilla it is a suggestion to change the handler return types to [return Response objects](https://starlite-api.github.io/starlite/usage/5-responses/3-returning-responses/). Check out that part of the docs, see if you agree or not, and lets start talking with a PR.

**Comment by ingjavierpinilla** [comment]
> 

Sounds good to me, should we implement this change in the return types and the matter of this issue in the same PR or should we keep these two implementations in separate PRs?

**Comment by RSPersonal** [comment]
@ingjavierpinilla which part would you like to pickup ?

**Comment by peterschutt** [comment]
> > 
> 
> Sounds good to me, should we implement this change in the return types and the matter of this issue in the same PR or should we keep these two implementations in separate PRs?

Fine to roll them together :+1:

**Comment by peterschutt** [comment]
closed in #430
